compositor_render/src/transformations/web_renderer.rs

```diff
@@ -1,86 +1,94 @@
+impl Default for WebRendererOptions {
+    fn default() -> Self {
+        Self {
+            init: true,
+            disable_gpu: false,
+        }
+    }
+}
 
 pub struct WebRenderer {
-    session_id: SessionId,
     #[allow(dead_code)]
     params: WebRendererTransformationParams,
+    // NOTE: Will be used for accessing V8 context later
+    _browser: cef::Browser,
+    state: Mutex<BrowserState>,
+    bgra_texture: BGRATexture,
+    bgra_bind_group: wgpu::BindGroup,
 }
 
 impl WebRenderer {
     pub fn new(
         ctx: &RegisterTransformationCtx,
         params: WebRendererTransformationParams,
     ) -> Result<Self, WebRendererNewError> {
-        // TODO: wait electron api by checking tcp connection on that port
-        thread::sleep(Duration::from_secs(5));
-        let session_id = ctx
-            .electron
-            .client
-            .new_session(&params.url, params.resolution)?;
-
-        Ok(Self { session_id, params })
+        info!("Starting web renderer for {}", &params.url);
+
+        let (painted_frames_sender, painted_frames_receiver) = crossbeam_channel::bounded(1);
+        let state = Mutex::new(BrowserState::new(painted_frames_receiver));
+        let client = BrowserClient::new(painted_frames_sender, params.resolution);
+        let browser = ctx.chromium.start_browser(&params.url, client)?;
+
+        let bgra_texture = BGRATexture::new(&ctx.wgpu_ctx, params.resolution);
+        let bgra_bind_group =
+            bgra_texture.new_bind_group(&ctx.wgpu_ctx, &ctx.wgpu_ctx.bgra_bind_group_layout);
+
+        Ok(Self {
+            params,
+            _browser: browser,
+            state,
+            bgra_texture,
+            bgra_bind_group,
+        })
     }
 
     pub fn render(
         &self,
         ctx: &RenderCtx,
         _sources: &[(&NodeId, &NodeTexture)],
         target: &NodeTexture,
-    ) -> Result<(), WebRendererRenderError> {
-        let frame = ctx.electron.client.get_frame(&self.session_id)?;
-        if !frame.is_empty() {
-            Self::write_texture(ctx, &frame, &target.rgba_texture())?;
-        }
-
-        Ok(())
-    }
+    ) {
+        let mut state = self.state.lock().unwrap();
+        let frame = state.retrieve_frame();
 
-    fn write_texture(
-        ctx: &RenderCtx,
-        data: &[u8],
-        target: &RGBATexture,
-    ) -> Result<(), WebRendererRenderError> {
-        let size = target.size();
-        let img = image::load_from_memory(data)?;
-
-        if img.width() != size.width || img.height() != size.height {
-            return Err(WebRendererRenderError::InvalidFrameResolution {
-                expected: Resolution {
-                    width: size.width as usize,
-                    height: size.height as usize,
-                },
-                received: Resolution {
-                    width: img.width() as usize,
-                    height: img.height() as usize,
-                },
-            });
+        if !frame.is_empty() {
```

Comment:
I know that it was like that before, but
- Why it would be empty?
- Can we refactor the code so it can't be empty?
- What would be rendered if this happens?
- If we can't do anything about it, at least use Option to signal when the value is missing instead of empty byte array

Reply:
On second thought, there is a chance that chromium would take longer time to startup a website so it's possible that compositor requests a render when the frame is not yet ready.